src/cart/src/cart/crt_context.c

```diff
@@ -1293,37 +1261,94 @@ crt_progress(crt_context_t crt_ctx, int64_t timeout,
 			hg_timeout = timeout;
 	}
 
-	while (true) {
-		crt_context_timeout_check(ctx);
-		crt_exec_progress_cb(ctx);
+	/** loop until callback returns non-null value */
+	while ((rc = cond_cb(arg)) == 0) {
+		int ret;
 
+		/**
+		 * Call progress once before processing timeouts in case
+		 * in case any replies are pending in the queue
+		 */
 		rc = crt_hg_progress(&ctx->cc_hg_ctx, hg_timeout);
```

Comment:
For this crt_hg_progress() before crt_context_timeout_check()/crt_exec_progress_cb(), seems better to pass hg_timeout as zero?

Reply:
good catch!